src/lib.rs

```diff
@@ -80,54 +83,57 @@ where
-                        let _result = open_sender.send(result);
+        self.event_loop
+            .run(move |event, target, control_flow| {
+                *control_flow = ControlFlow::Wait;
+                match event {
+                    Event::WindowEvent { window_id, event } => {
+                        let event = WindowEvent::from(event);
+                        self.running
+                            .windows
+                            .send(window_id, WindowMessage::Event(event));
                     }
-                    EventLoopMessage::User {
-                        message,
-                        response_sender,
-                    } => {
-                        let _result = response_sender
-                            .send((self.message_callback)(message, &self.running.windows));
+                    Event::RedrawRequested(window_id) => {
+                        self.running.windows.send(window_id, WindowMessage::Redraw);
                     }
-                },
-                Event::NewEvents(_)
-                | Event::DeviceEvent { .. }
-                | Event::Suspended
-                | Event::Resumed
-                | Event::MainEventsCleared
-                | Event::RedrawEventsCleared
-                | Event::LoopDestroyed => {}
-            }
-        });
+                    Event::UserEvent(message) => match message {
+                        EventLoopMessage::CloseWindow(window_id) => {
+                            if self.running.windows.close(window_id) {
+                                *control_flow = ControlFlow::ExitWithCode(0);
+                            }
+                        }
+                        EventLoopMessage::WindowPanic(window_id) => {
+                            if self.running.windows.close(window_id) {
+                                *control_flow = ControlFlow::ExitWithCode(1);
+                            }
+                        }
+                        EventLoopMessage::OpenWindow {
+                            attrs,
+                            sender,
+                            open_sender,
+                        } => {
+                            let result = self.running.windows.open(target, attrs, sender);
+                            let _result = open_sender.send(result);
+                        }
+                        EventLoopMessage::User {
+                            message,
+                            response_sender,
+                        } => {
+                            let _result = response_sender
+                                .send((self.message_callback)(message, &self.running.windows));
+                        }
+                    },
+                    Event::NewEvents(_)
+                    | Event::DeviceEvent { .. }
+                    | Event::Suspended
+                    | Event::Resumed
+                    | Event::LoopExiting
+                    | Event::AboutToWait => {}
```

Comment:
These two are new

Reply:
run() should mimic winit's run(). If it now returns a result, we should too.